<?xml version="1.0" ?>
<robot name="arm_bot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- This is all in cm and radians -->
  <xacro:property name="l1" value="10" />
  <xacro:property name="l2" value="12" />
  <xacro:property name="base_height" value="7.5" />
  <xacro:property name="zero" value="1e-12" />
  <xacro:property name="link_w" value="1.5" />
  <xacro:property name="link_h" value="1.5" />

  <link name="world"/>
  <link name="end_effector_link"/>

  <link name="base1">
    <visual>
        <geometry>
            <box size="${zero} ${zero} ${zero}" />
        </geometry>
        <material name="blue">
            <color rgba="0 0 .8 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <box size="${zero} ${zero} ${zero}" />
        </geometry>
    </collision>
    
    <inertial>
        <mass value="0.1" />
        <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0.0" ixz="0.0" iyz="0.0" />
    </inertial>
  </link>

  <link name="base2">
    <visual>
        <origin rpy="0 0 0" xyz="0 0 ${base_height/2}" />
        <geometry>
            <box size="${link_w} ${link_h} ${base_height}" />
        </geometry>
        <material name="blue">
            <color rgba="0 0 .8 1" />
        </material>
    </visual>
    <collision>
        <origin rpy="0 0 0" xyz="0 0 ${base_height/2}" />
        <geometry>
            <box size="${link_w} ${link_h} ${base_height}" />
        </geometry>
    </collision>
    
    <inertial>
        <mass value="0.1" />
        <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0.0" ixz="0.0" iyz="0.0" />
    </inertial>
  </link>

  <link name="link1">
    <visual>
        <origin rpy="0 0 0" xyz="0 ${l1/2} 0" />
        <geometry>
            <box size="${link_w} ${l1} ${link_h}" />
        </geometry>
        <material name="blue">
            <color rgba="0 0 .8 1" />
        </material>
    </visual>
    <collision>
        <origin rpy="0 0 0" xyz="0 ${l1/2} 0" />
        <geometry>
            <box size="${link_w} ${l1} ${link_h}" />
        </geometry>
    </collision>
    
    <inertial>
        <mass value="0.1" />
        <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0.0" ixz="0.0" iyz="0.0" />
    </inertial>
  </link>

  <link name="link2">
    <visual>
        <origin rpy="0 0 0" xyz="0 ${l2/2} 0" />
        <geometry>
            <box size="${link_w} ${l2} ${link_h}" />
        </geometry>
        <material name="blue">
            <color rgba="0 0 .8 1" />
        </material>
    </visual>
    <collision>
        <origin rpy="0 0 0" xyz="0 ${l2/2} 0" />
        <geometry>
            <box size="${link_w} ${l2} ${link_h}" />
        </geometry>
    </collision>
    
    <inertial>
        <mass value="0.1" />
        <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0.0" ixz="0.0" iyz="0.0" />
    </inertial>
  </link>

  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base1"/>
  </joint>
  <joint name="base_rotate" type="continuous">
      <parent link="base1" />
      <child link="base2" />
      <origin rpy="0 0 0" xyz="0 0 0" />
      <axis xyz="0 0 1" />
  </joint>
  <joint name="joint1" type="revolute">
      <parent link="base2" />
      <child link="link1" />
      <origin rpy="0 0 0" xyz="0 0 ${base_height}" />
      <axis xyz="1 0 0" />
      <limit lower="-3.14" upper="3.14" effort="100" velocity="3" />
      <!-- Effort is between 0 and 100 (units are %) and velocity is radians per second -->
  </joint>
  <joint name="joint2" type="revolute">
    <parent link="link1" />
    <child link="link2" />
    <origin rpy="0 0 0" xyz="0 ${l1} 0" />
    <axis xyz="1 0 0" />
    <limit lower="-3.14" upper="3.14" effort="100" velocity="3"/>
    <!-- Effort is between 0 and 100 (units are %) and velocity is radians per second -->
  </joint>
  <joint name="end_effector_joint" type="fixed">
      <parent link="link2" />
      <child link="end_effector_link" />
      <origin rpy="0 0 0" xyz="0 ${l2} 0" />
  </joint>
  
  <transmission name="trans_base_rotate">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="base_rotate">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="base_rotate_motor">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
  </transmission>
  <transmission name="trans_joint1">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="joint1">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="joint1_motor">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
  </transmission>
  <transmission name="trans_joint2">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="joint2">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="joint2_motor">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
  </transmission>
  <gazebo>
      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
          <robotNamespace>/</robotNamespace>
      </plugin>
  </gazebo>
</robot>